cc,wasi: support WASI reactors via -mexec-model flag

mathetake opened the pull request:
Based on this patch to LLVM (https://reviews.llvm.org/D62922). This partially resolves (only for zig cc) #6757

kubkon commented on src/wasi_libc.zig:
```diff
@@ -51,16 +51,20 @@ pub fn buildWasiLibcSysroot(comp: *Compilation) !void {
 
-        var comp_sources = std.ArrayList(Compilation.CSourceFile).init(arena);
         for (crt_src_files) |file_path| {
+            var comp_sources = std.ArrayList(Compilation.CSourceFile).init(arena);
             try comp_sources.append(.{
                 .src_path = try comp.zig_lib_directory.join(arena, &[_][]const u8{
                     "libc", try sanitize(arena, file_path),
                 }),
                 .extra_flags = args.items,
             });
+            try comp.build_crt_file(
+                file_path[std.mem.lastIndexOf(u8, file_path, "/").? + 1 .. std.mem.lastIndexOf(u8, file_path, ".").?],
+                .Obj,
+                comp_sources.items,
+            );
```
These changes are no longer necessary and are fixed in https://github.com/ziglang/zig/pull/8992

mathetake replied:
sounds good! so should I wait until that PR is merged?

kubkon replied:
Actually, feel free to rebase onto my branch, and we can carry on from there. Feel free to ping me on Discord if you have any questions about the changes in that PR or whatnot!

kubkon reviewed (changes requested):
Thanks for the PR! I've got a couple of suggestions that when we address, I'm sure we'll be able to merge this change no problem.

mathetake commented on src/Compilation.zig:
```diff
@@ -719,6 +719,7 @@ pub const InitOptions = struct {
     test_filter: ?[]const u8 = null,
     test_name_prefix: ?[]const u8 = null,
     subsystem: ?std.Target.SubSystem = null,
+    want_reactor_exec_model: bool = false,
```
Thanks, that would be much better! Done in https://github.com/ziglang/zig/pull/8978/commits/6614a204d060411b4bbbdbcd035ccee919a5c006

mathetake commented:
OK so I verified this works as expected locally.

mathetake commented:
@kubkon ok now rebased and added a necessary commit for checking in command/reactor crt (https://github.com/ziglang/zig/pull/8978/commits/a2b20449efdb07ba28b739fa544d6364d6a5851c). PTAL!

kubkon commented on src/Compilation.zig:
```diff
@@ -1444,6 +1451,8 @@ pub fn create(gpa: *Allocator, options: InitOptions) !*Compilation {
             // TODO add logic deciding which crt1 we want here.
             comp.work_queue.writeAssumeCapacity(&[_]Job{
                 .{ .wasi_libc_crt_file = .crt1_o },
+                .{ .wasi_libc_crt_file = .crt1_command_o },
+                .{ .wasi_libc_crt_file = .crt1_reactor_o },
```
OK, so my idea for this bit here is similar to the logic just a few lines above we use for working out which emulated WASI libc subcomponents we should build in this run (unless they are already in global cache). So, instead of building _all_ CRTs, how about we now use the new option you just added to work out precisely which CRT we should include in our build, something like:

```zig
const crt_file: wasi_libc.CRTFile = if (comp.bin_file.options.wasi_exec_model) |exec_model| crt_file: {
    switch (exec_model) {
        .command => break :crt_file .crt1_command_o,
        .reactor => break :crt_file .crt1_reactor_o,
    }
} else .crt1_o;
comp.work_queue.writeAssumeCapacity(&[_]Job{
    .{ .wasi_libc_crt_file = crt_file },
    .{ .wasi_libc_crt_file = .libc_a },
});
```

kubkon replied:
Also, while we're here, could you remove the TODO as it is no longer needed.

mathetake replied:
fixed in https://github.com/ziglang/zig/pull/8978/commits/cd610975f2c401ec413787c04c2c1c8a34e54804 with a slight modification. I couldn't compile without explicitly giving the enum value inside of these switch branches (maybe due to type inference limitation)?

mathetake replied:
Anyway thank you so much for suggestions!

kubkon commented on src/link.zig:
```diff
@@ -117,6 +117,8 @@ pub const Options = struct {
     version: ?std.builtin.Version,
     libc_installation: ?*const LibCInstallation,
 
+    wasi_exec_model: ?Compilation.WasiExecModel = null,
```
Could you add a sentence or two explaining what this option holds?

mathetake replied:
https://github.com/ziglang/zig/pull/8978/commits/cd610975f2c401ec413787c04c2c1c8a34e54804 fixed!

kubkon commented on src/Compilation.zig:
```diff
@@ -725,6 +730,7 @@ pub const InitOptions = struct {
     test_filter: ?[]const u8 = null,
     test_name_prefix: ?[]const u8 = null,
     subsystem: ?std.Target.SubSystem = null,
+    wasi_exec_model: ?WasiExecModel = null,
```
Could you add a sentence or two explaining what this option holds?

mathetake replied:
https://github.com/ziglang/zig/pull/8978/commits/cd610975f2c401ec413787c04c2c1c8a34e54804 fixed!

kubkon commented on src/main.zig:
```diff
@@ -1254,6 +1255,14 @@ fn buildOutputType(
+                        else if (std.mem.eql(u8, it.only_arg, "command"))
+                            Compilation.WasiExecModel.command
+                        else
+                            null;
```
Minor nit: you actually don't need to initialize to `null` here since it is already initialized to this value by default.

mathetake replied:
fixed in  https://github.com/ziglang/zig/pull/8978/commits/230878352006e466f66600a979bdba8bddc98acd

kubkon commented on src/link/Wasm.zig:
I think it is more readable to unpack the optional using the `if (opt) |value|` syntax which is more readable, and use labeled block to leave early if the inner value is not a `.reactor`:

```zig
if (self.base.options.wasi_exec_model) |exec_model| blk: {
    if (exec_model != .reactor) break :blk;
    try argv.append("--no-entry");
}
```

See [Optionals](https://ziglang.org/documentation/0.8.0/#toc-Optionals) for more details.

kubkon reviewed (approved):
Looks great! Thanks for working on this @mathetake!

kubkon commented:
BTW once #8992 lands, I'll merge this one next.

mathetake commented:
@kubkon Thanks for the review! Much appreciated, and I learned a lot from your suggestions:)

kubkon commented:
OK Github, why the heck did you close this PR?!

kubkon commented:
@mathetake would you mind reopening the PR?

kubkon commented:
Or I can try merging your branch manually on the command line, let's see...

mathetake commented:
OK, let me reopen!

kubkon commented:
> OK, let me reopen!

Don't worry about conflicts, I'll take care of them while merging.

mathetake commented:
https://github.com/ziglang/zig/pull/9052 here you go!
